mpc/README.md

```diff
@@ -81,9 +84,10 @@ Digest written to: response.bin.digest
 ...
 ```
 
-For phase2:
+For phase2 (assuming `mpc-client-phase2` is in the `PATH`):
 ```console
-(env) $ phase2_contribute https://<host>[:<port>] contributor.key
+(env) $ phase2_contribute https://<host>[:<port>] contributor.key \
+            --mpc-tool $(which mpc-client-phase2)
```

Comment:
(related to the above, we should be able to remove the need for this flag in most cases)

Reply:
The current code looks into default locations that are relative paths from the "contributor"'s repo. This makes the assumptions that:
1. the contributor compiled the binaries from source (which can be desirable actually - to verify the code the ends up being ran), and
2. has cloned `powersoftau` and `zeth` under the same parent directory. Any other structure will break since the tool will look in the wrong location.

If we want to package everything in docker images, and use a multi-stage build to build the binaries and move them in the final image, then we will get rid of all the useless code and just move binaries around, which won't be under `zeth/build/mpc_tools/mpc_phase2/`, same for PoT binaries which won't be under `/powersoftau/target/release`, hence the flag is quite helpful here.
Actually anyone running the `setup_env.sh` script will already have the Zeth binaries in is PATH, and thus should be able to run this command without any trouble.

I can remove the `--mpc-tool $(which mpc-client-phase2)` here and put `--mpc-tool <path-to-mpc-binary>`, that works for me either way - the second alternative just forces the user to look for the path to the MPC binary and paste it here.
Else, we need to modify:
`(Phase1 only) clone and build: https://github.com/clearmatics/powersoftau` to rather say: `(Phase1 only) clone and build: https://github.com/clearmatics/powersoftau under the same parent directory as Zeth` to make sure the default path in `powersofau_command.py` (i.e. `../../../powersoftau/target/<CONFIG>`) resolves to the PoT binaries which is not enforced here.

Reply:
> The current code looks into default locations that are relative paths from the "contributor"'s repo. This makes the assumptions that:
> 
>     1. the contributor compiled the binaries from source (which can be desirable actually - to verify the code the ends up being ran), and
> 
>     2. has cloned `powersoftau` and `zeth` under the same parent directory. Any other structure will break since the tool will look in the wrong location.

Yes, exactly.  `powersoftau` might be a slightly different case since it's external, but for the `--mpc-tool` flag here, I'd expect binary to be in a "known" location if the user is using defaults.  If the user is doing something custom, that's exactly what the flag is for.

> If we want to package everything in docker images, and use a multi-stage build to build the binaries and move them in the final image, then we will get rid of all the useless code and just move binaries around, which won't be under `zeth/build/mpc_tools/mpc_phase2/`, same for PoT binaries which won't be under `/powersoftau/target/release`, hence the flag is quite helpful here.
> Actually anyone running the `setup_env.sh` script will already have the Zeth binaries in is PATH, and thus should be able to run this command without any trouble.

Ah, good point.  I usually keep a plain env so I forgot about the `setup_env.sh` script.  So that's true that the command will be in the path if it has been built locally in the default location (your assumption 1 above) and the user has run `setup_env.py`.  However, if setup_env.sh *has* been run, it seems like even more reason not to require the user to specify the location again here (if I understood correctly, that's part of the intention behind `setup_env.sh`).

In the case of a docker image, I'd expect the binaries to be placed somewhere convenient so that again the user doesn't have to give this flag (or we'd wrap everything in a script).  

> I can remove the `--mpc-tool $(which mpc-client-phase2)` here and put `--mpc-tool <path-to-mpc-binary>`, that works for me either way - the second alternative just forces the user to look for the path to the MPC binary and paste it here.
> Else, we need to modify:
> `(Phase1 only) clone and build: https://github.com/clearmatics/powersoftau` to rather say: `(Phase1 only) clone and build: https://github.com/clearmatics/powersoftau under the same parent directory as Zeth` to make sure the default path in `powersofau_command.py` (i.e. `../../../powersoftau/target/<CONFIG>`) resolves to the PoT binaries which is not enforced here.

I was thinking more of the mpc case, since we have more control over that, but in general I tend to aim for an approach where all commands are as simple as possible if the user has followed the README.  If the user has done something custom, we add hints such as: 
```
note: if mpc tools have been built or installed into a directory other than `build` in the repo root, the location must be specified with the `--mpc-tool` flag.  See `--help` for more details.
```

So, for the `powersoftau` location (which is a definitely more painful), your suggestion seems sensible to me: give a recommended install location (maybe with a note why), and "otherwise the location can be specified manually ... see --help".

(Eventually, we might consider making pwoersoftau a submodule, and have the python wrapper build it on demand, but that's far out of scope here, of course.)

Reply:
>  However, if setup_env.sh has been run, it seems like even more reason not to require the user to specify the location again here (if I understood correctly, that's part of the intention behind setup_env.sh).

Exact, that's why I used `$(which CMD)` as a way to directly get the path to the command (and not requiring the user to specify it).

> In the case of a docker image, I'd expect the binaries to be placed somewhere convenient

Yes I moved all the binaries in the PATH by default (see: Docker PR for more details). So users can copy and paste:
```
phase2_contribute https://mpc-contributor:<port> contributor.key \
            --mpc-tool $(which mpc-client-phase2)
```
and it will just work.

> so that again the user doesn't have to give this flag

How would it be possible with the current code base? The default path is `../../build/mpc_tools/mpc_phase2/<tool>`, so to save the flag from the command, one would need to build nested folders that reflect the relative path above (same for PoT btw), no?

Reply:
> Ah, good point. I usually keep a plain env so I forgot about the setup_env.sh script. So that's true that the command will be in the path if it has been built locally in the default location (your assumption 1 above) and the user has run setup_env.py.

Note that instead of modifying the PATH, we can also move the binaries to `/usr/local/bin` (move them in the PATH) with a `make install` rule. As long as they are in the PATH, we can locate them with `which`. 
We could actually "wrap" the `which` command to locate the binaries, didn't aim to change too many things here but that would remove the need for the flag (which is probably what you had in mind?)..

Reply:
As discussed off-line the best approach seems to be to mimic what we've done on the cpp side, where we use different levels of priority between flags, config files and env vars. Likewise, we can use env vars here to point to the binaries, and fall back to the path that is currently hardcoded in the code if the env var is not defined.

Reply:
Just to complete the discussion, it sounds like the original issue was with trying to manually move binaries out of the build directory.
This case has not been addressed (the default build does not support this and the assumption was that users would leave the binaries in the default build location), and hence the `--mpc-tool` flag was required in these circumstances.

The right way to address this would be to create some kind of install process, so that the destination is specified once and all tools can be installed with the information they need to find each other.

For the ad-hoc case of the binaries in arbitrary locations, an env-var can be used as an alternative to this flag.  This would still rely on several assumptions (like the binaries being statically linked or being able to find their dynamic dependencies), but could be useful for Docker contrainers where the state of the system can be preconfigured.

But ideally, for deployment, we would support:
```
$ cmake [CMAKE_INSTALL_PREFIX=/usr/local]
$ make install
```

Since the tools are generally only intended to be used via the python wrappers, it could be really nice to encapsulate everything in a python pacakge (so the native tools are build at `pip install` time, as with other packages).  This would give uses a one-shot install command.